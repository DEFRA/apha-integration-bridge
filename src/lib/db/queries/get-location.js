import Joi from 'joi'
import { query } from '../operations/query.js'

export const GetLocationSchema = Joi.object({
  locationId: Joi.string()
    .trim()
    .pattern(/^L\d+$/)
    .required()
    .description('Location ID (e.g., L97339)')
})

/**
 * @typedef {Record<string, Array<(value: unknown) => unknown>>} Marshallers
 * @typedef {{ sql: string; bindings: readonly unknown[]; marshallers?: Marshallers }} Query
 */
export function getLocation(locationId) {
  const { value, error } = GetLocationSchema.validate({ locationId })
  if (error) {
    throw new Error(`Invalid parameters: ${error.message}`)
  }

  const knex = query()

  // Shared column list (keep aliases identical across both branches)
  const sharedSelect = {
    locationId: 'loc.location_id',
    paonStartNumber: 'ba.paon_start_number',
    paonStartNumberSuffix: 'ba.paon_start_number_suffix',
    paonEndNumber: 'ba.paon_end_number',
    paonEndNumberSuffix: 'ba.paon_end_number_suffix',
    paonDescription: 'ba.paon_description',
    saonDescription: 'ba.saon_description',
    saonStartNumber: 'ba.saon_start_number',
    saonStartNumberSuffix: 'ba.saon_start_number_suffix',
    saonEndNumber: 'ba.saon_end_number',
    saonEndNumberSuffix: 'ba.saon_end_number_suffix',
    street: 'ba.street',
    locality: 'ba.locality',
    town: 'ba.town',
    administrativeAreaCounty: 'ba.administrative_area',
    postcode: 'ba.postcode',
    ukInternalCode: 'ba.uk_internal_code',
    countryCode: 'ba.country_code'
  }

  // Branch 1: Livestock Units
  const luQb = knex
    .from({ loc: 'ahbrp.location' })
    .join({ fa: 'ahbrp.feature_address' }, 'loc.feature_pk', 'fa.feature_pk')
    .join({ fs: 'ahbrp.feature_state' }, 'loc.feature_pk', 'fs.feature_pk')
    .join({ ba: 'ahbrp.bs7666_address' }, 'fa.address_pk', 'ba.address_pk')
    .join({ al: 'ahbrp.asset_location' }, 'loc.feature_pk', 'al.feature_pk')
    .join({ lu: 'ahbrp.livestock_unit' }, 'al.asset_pk', 'lu.asset_pk')
    .join({ ass: 'ahbrp.asset_state' }, 'lu.asset_pk', 'ass.asset_pk')
    .whereNull('fa.feature_address_to_date')
    .whereNull('al.asset_location_to_date')
    .where('al.asset_location_type', 'PRIMARYLOCATION')
    .where('ass.asset_status_code', '<>', 'INACTIVE')
    .whereNull('ass.asset_state_to_dttm')
    .where('fs.feature_status_code', '<>', 'INACTIVE')
    .whereNull('fs.feature_state_to_dttm')
    .where('loc.location_id', value.locationId)
    .select(sharedSelect)
    // Cast to common type to satisfy UNION typing
    .select(knex.raw('CAST(?? AS VARCHAR2(50)) "unitId"', ['lu.unit_id']))
    .select(knex.raw(`'LU' as "unitType"`))

  // Branch 2: Facilities
  const facQb = knex
    .from({ loc: 'ahbrp.location' })
    .join({ fa: 'ahbrp.feature_address' }, 'loc.feature_pk', 'fa.feature_pk')
    .join({ fs: 'ahbrp.feature_state' }, 'loc.feature_pk', 'fs.feature_pk')
    .join({ ba: 'ahbrp.bs7666_address' }, 'fa.address_pk', 'ba.address_pk')
    .join({ al: 'ahbrp.asset_location' }, 'loc.feature_pk', 'al.feature_pk')
    .join({ facility: 'ahbrp.facility' }, 'al.asset_pk', 'facility.asset_pk')
    .join({ ass: 'ahbrp.asset_state' }, 'facility.asset_pk', 'ass.asset_pk')
    .whereNull('fa.feature_address_to_date')
    .whereNull('al.asset_location_to_date')
    .where('al.asset_location_type', 'PRIMARYLOCATION')
    .where('ass.asset_status_code', '<>', 'INACTIVE')
    .whereNull('ass.asset_state_to_dttm')
    .where('fs.feature_status_code', '<>', 'INACTIVE')
    .whereNull('fs.feature_state_to_dttm')
    .where('loc.location_id', value.locationId)
    .select(sharedSelect)
    // Cast here too; if FACILITY.UNIT_ID is numeric in prod, this is essential
    .select(knex.raw('CAST(?? AS VARCHAR2(50)) "unitId"', ['facility.unit_id']))
    .select(knex.raw(`'F' as "unitType"`))

  const qb = luQb.union([facQb]) // UNION (not ALL) as per your SQL

  const { sql, bindings } = qb.toSQL()
  return { sql, bindings }
}
