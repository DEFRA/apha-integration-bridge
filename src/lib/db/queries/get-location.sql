SELECT
LOCATION_ID,
PAON_START_NUMBER paonstartnumber,
PAON_START_NUMBER_SUFFIX paonstartnumbersuffix,
PAON_END_NUMBER paonendnumber,
PAON_END_NUMBER_SUFFIX paonendnumbersuffix,
PAON_DESCRIPTION paondescription,
SAON_DESCRIPTION saondescription,
SAON_START_NUMBER saonstartnumber,
SAON_START_NUMBER_SUFFIX saonstartnumbersuffix,
SAON_END_NUMBER saonendnumber,
SAON_END_NUMBER_SUFFIX saonendnumbersuffix,
STREET,
LOCALITY,
TOWN,
ADMINISTRATIVE_AREA COUNTY,
POSTCODE,
UK_INTERNAL_CODE ukinternalcode,
COUNTRY_CODE countrycode,
UNIT_ID unitid,
'LU' unittype

FROM
AHBRP.LOCATION,
AHBRP.FEATURE_ADDRESS FA,
AHBRP.FEATURE_STATE FS,
AHBRP.BS7666_ADDRESS BA,
AHBRP.ASSET_LOCATION AL,
AHBRP.LIVESTOCK_UNIT LU,
AHBRP.ASSET_STATE ASS

WHERE
LOCATION.FEATURE_PK = FA.FEATURE_PK
AND 
LOCATION.FEATURE_PK = FS.FEATURE_PK
AND
FA.ADDRESS_PK = BA.ADDRESS_PK
AND
LOCATION.FEATURE_PK = AL.FEATURE_PK
AND
AL.ASSET_PK = LU.ASSET_PK
AND
LU.ASSET_PK = ASS.ASSET_PK
AND
FA.FEATURE_ADDRESS_TO_DATE IS NULL
AND
AL.ASSET_LOCATION_TO_DATE IS NULL
AND
AL.ASSET_LOCATION_TYPE ='PRIMARYLOCATION'
AND
ASS.ASSET_STATUS_CODE <> 'INACTIVE'
AND
ASS.ASSET_STATE_TO_DTTM IS NULL
AND
FS.FEATURE_STATUS_CODE <> 'INACTIVE'
AND 
FS.FEATURE_STATE_TO_DTTM IS NULL
AND
LOCATION.LOCATION_ID ='L97339'

union

SELECT
LOCATION_ID,
PAON_START_NUMBER,
PAON_START_NUMBER_SUFFIX,
PAON_END_NUMBER,
PAON_END_NUMBER_SUFFIX,
PAON_DESCRIPTION,
SAON_DESCRIPTION,
SAON_START_NUMBER,
SAON_START_NUMBER_SUFFIX,
SAON_END_NUMBER,
SAON_END_NUMBER_SUFFIX,
STREET,
LOCALITY,
TOWN,
ADMINISTRATIVE_AREA COUNTY,
POSTCODE,
UK_INTERNAL_CODE,
COUNTRY_CODE,
UNIT_ID,
'F' UNIT_TYPE

FROM
AHBRP.LOCATION,
AHBRP.FEATURE_ADDRESS FA,
AHBRP.FEATURE_STATE FS,
AHBRP.BS7666_ADDRESS BA,
AHBRP.ASSET_LOCATION AL,
AHBRP.FACILITY,
AHBRP.ASSET_STATE ASS

WHERE
LOCATION.FEATURE_PK = FA.FEATURE_PK
AND 
LOCATION.FEATURE_PK = FS.FEATURE_PK
AND
FA.ADDRESS_PK = BA.ADDRESS_PK
AND
LOCATION.FEATURE_PK = AL.FEATURE_PK
AND
AL.ASSET_PK = FACILITY.ASSET_PK
AND
FACILITY.ASSET_PK = ASS.ASSET_PK
AND
FA.FEATURE_ADDRESS_TO_DATE IS NULL
AND
AL.ASSET_LOCATION_TO_DATE IS NULL
AND
AL.ASSET_LOCATION_TYPE='PRIMARYLOCATION'
AND
ASS.ASSET_STATUS_CODE<> 'INACTIVE'
AND
ASS.ASSET_STATE_TO_DTTM IS NULL
AND
FS.FEATURE_STATUS_CODE <> 'INACTIVE'
AND
FS.FEATURE_STATE_TO_DTTM IS NULL
AND
LOCATION.LOCATION_ID= :location
