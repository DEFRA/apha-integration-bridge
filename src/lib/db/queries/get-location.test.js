import { test, expect } from '@jest/globals'

import { getLocation } from './get-location.js'

test('returns the expected query for valid parameters', () => {
  const locationId = 'L97339'

  const { sql, bindings } = getLocation(locationId)

  expect(sql).toMatchSnapshot()

  expect(bindings).toEqual({
    location: locationId
  })
})

test('throws if the parameters are invalid', () => {
  expect(() => getLocation(null)).toThrow(/invalid/i)
})
