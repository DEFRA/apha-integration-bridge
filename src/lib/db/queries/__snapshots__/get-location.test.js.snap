// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns the expected query for valid parameters 1`] = `"select loc.location_id locationId, ba.paon_start_number paonStartNumber, ba.paon_start_number_suffix paonStartNumberSuffix, ba.paon_end_number paonEndNumber, ba.paon_end_number_suffix paonEndNumberSuffix, ba.paon_description paonDescription, ba.saon_description saonDescription, ba.saon_start_number saonStartNumber, ba.saon_start_number_suffix saonStartNumberSuffix, ba.saon_end_number saonEndNumber, ba.saon_end_number_suffix saonEndNumberSuffix, ba.street street, ba.locality locality, ba.town town, ba.administrative_area administrativeAreaCounty, ba.postcode postcode, ba.uk_internal_code ukInternalCode, ba.country_code countryCode, CAST(lu.unit_id AS VARCHAR2(50)) "unitId", 'LU' as "unitType" from ahbrp.location loc inner join ahbrp.feature_address fa on loc.feature_pk = fa.feature_pk inner join ahbrp.feature_state fs on loc.feature_pk = fs.feature_pk inner join ahbrp.bs7666_address ba on fa.address_pk = ba.address_pk inner join ahbrp.asset_location al on loc.feature_pk = al.feature_pk inner join ahbrp.livestock_unit lu on al.asset_pk = lu.asset_pk inner join ahbrp.asset_state ass on lu.asset_pk = ass.asset_pk where fa.feature_address_to_date is null and al.asset_location_to_date is null and al.asset_location_type = ? and ass.asset_status_code <> ? and ass.asset_state_to_dttm is null and fs.feature_status_code <> ? and fs.feature_state_to_dttm is null and loc.location_id = ? union select loc.location_id locationId, ba.paon_start_number paonStartNumber, ba.paon_start_number_suffix paonStartNumberSuffix, ba.paon_end_number paonEndNumber, ba.paon_end_number_suffix paonEndNumberSuffix, ba.paon_description paonDescription, ba.saon_description saonDescription, ba.saon_start_number saonStartNumber, ba.saon_start_number_suffix saonStartNumberSuffix, ba.saon_end_number saonEndNumber, ba.saon_end_number_suffix saonEndNumberSuffix, ba.street street, ba.locality locality, ba.town town, ba.administrative_area administrativeAreaCounty, ba.postcode postcode, ba.uk_internal_code ukInternalCode, ba.country_code countryCode, CAST(facility.unit_id AS VARCHAR2(50)) "unitId", 'F' as "unitType" from ahbrp.location loc inner join ahbrp.feature_address fa on loc.feature_pk = fa.feature_pk inner join ahbrp.feature_state fs on loc.feature_pk = fs.feature_pk inner join ahbrp.bs7666_address ba on fa.address_pk = ba.address_pk inner join ahbrp.asset_location al on loc.feature_pk = al.feature_pk inner join ahbrp.facility facility on al.asset_pk = facility.asset_pk inner join ahbrp.asset_state ass on facility.asset_pk = ass.asset_pk where fa.feature_address_to_date is null and al.asset_location_to_date is null and al.asset_location_type = ? and ass.asset_status_code <> ? and ass.asset_state_to_dttm is null and fs.feature_status_code <> ? and fs.feature_state_to_dttm is null and loc.location_id = ?"`;
